SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp

```diff
@@ -0,0 +1,204 @@
+    // for fill.
+    for (size_t i = offset; i < Stride * NumSelectedElems + offset;
+         i += Stride) {
+      DataT expected = ref_data_for_fill[(i - offset) / Stride];
+      DataT retrieved = result[i];
+      if (expected != retrieved) {
+        passed = fail_test(i, expected, retrieved, data_type);
+      }
+    }
+    // Verify that values, that do not was selected has initial values.
+    for (size_t i = 0; i < NumElems; ++i) {
+      if (i % Stride != 0) {
+        DataT expected = initial_ref_data[i];
+        DataT retrieved = result[i];
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type);
+        }
+      }
+    }
+
+    if (!is_correct_type.value()) {
+      passed = false;
+      log::note("Test failed due to type of the object that returns " +
+                TestCaseT::get_description() +
+                " operator is not equal to the expected one for simd<" +
+                data_type + ", " + std::to_string(NumElems) + ">.");
+    }
+
+    return passed;
+  }
+
+private:
+  bool fail_test(size_t i, DataT expected, DataT retrieved,
+                 const std::string &data_type) {
+    const auto description = TestDescription<DataT, NumElems>(
+        i, retrieved, expected, NumSelectedElems, Stride, data_type);
+    log::fail(description);
+
+    return false;
+  }
+};
+
+template <tested_types TestedTypesT>
+bool run_test_with_chosen_data_types(sycl::queue &queue) {
+  bool passed = true;
+  constexpr int desired_simd_small_size = 1;
+  constexpr int desired_simd_large_size = 16;
+  constexpr int coefficient_of_division = 3;
+  constexpr int small_offset_value = 1;
+  const int large_offset_value =
+      desired_simd_large_size - std::ceil(2 * desired_simd_large_size / 3);
+
+  const auto small_size = get_dimensions<desired_simd_small_size>();
+  const auto great_size = get_dimensions<desired_simd_large_size>();
+  const auto all_types = get_tested_types<TestedTypesT>();
+
+  passed &=
+      for_all_combinations<run_test, select_rval, size<1>, stride<1>,
+                           do_not_use_offset>(all_types, small_size, queue, 0);
+  passed &=
+      for_all_combinations<run_test, select_rval, size<1>, stride<1>,
+                           do_not_use_offset>(all_types, small_size, queue, 0);
+  passed &= for_all_combinations<
+      run_test, select_rval,
+      size<desired_simd_large_size / coefficient_of_division>,
+      stride<coefficient_of_division>, do_not_use_offset>(all_types, great_size,
+                                                          queue, 0);
+  passed &= for_all_combinations<
+      run_test, select_rval,
+      size<desired_simd_large_size / coefficient_of_division>, stride<3>,
+      do_not_use_offset>(all_types, great_size, queue, 0);
+  passed &= for_all_combinations<
+      run_test, select_rval, size<3>,
+      stride<desired_simd_large_size / coefficient_of_division>,
+      do_not_use_offset>(all_types, great_size, queue, 0);
```

Comment:
What if we have something like this? Would it still work?

```suggestion
  passed &= for_all_combinations<
      run_test, select_rval, size<3>,
      stride<desired_simd_large_size / coefficient_of_division>,
      do_not_use_offset>(all_types, great_size, queue, 42);
```

Could we possibly avoid pairs of template parameter `use_offset`/`do_not_use_offset` and runtime parameter `offset` by using constexpr `offset` values?

Reply:
The `static_assert` has been added for this case in this commit: 179a8f24c9550018b59b5f10350c5b6a0c6b6faf